Fix spectator not starting from current player position

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/17267.

Changes to `PlayerLoader` are only required for testing. Can be split out into a separate PR if preferred, but I didn't bother because they are file (and commit) level isolated.

I think reading through the changes and attempting to understand from code is best (if it isn't 100% straightforward what has changed and why then I think further refactor may be required) so I'm not going to overly explain this one.

peppy committed:
Expose the loading player in `PlayerLoader`

peppy committed:
Add test coverage of `SpectatorPlayer` failing to seek on inopportune frame arrival time

peppy committed:
Use `nullable` in `PlayerLoader`

peppy committed:
Use a more explicit flow to set and reset `GameplayClockContainer` start time

peppy commented:
Note that I haven't touched the fact that int the first frame after `LoadComplete`, `Player` is not-present. I honestly don't know if that should be fixed or how to fix it (there's probably other screens in a similar scenario) so ensured the solution offered here doesn't rely on that part anymore.

frenzibyte commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -90,26 +89,33 @@ protected override void LoadComplete()
-                    startOffset = Math.Min(startOffset, firstHitObjectTime - beatmap.BeatmapInfo.AudioLeadIn);
-            }
+                // sane default provided by ruleset.
+                double offset = gameplayStartTime;
```
Would rename `gameplayStartTime` to not conflict with the introduced property `GameplayStartTime` (maybe `rulesetGameplayStartTime`).

peppy replied:
Wow, I didn't even see this variable conflict.

frenzibyte reviewed (commented):
Haven't deeply checked this, but the direction sounds good in general. 

`startAtGameplayStart` could potentially be removed in favour of `Player.SetGameplayStartTime`/`GameplayClockContainer.Reset(gameplayStartTime: ...)`.

peppy committed:
Tidy up constructor logic overlap with `gameplayStartTime`

peppy committed:
Rename start time calculation method and add more commenting to explain purpose better

peppy commented:
I've made quite serious changes as I misunderstood the constructor parameter initially (or missed it). Note that while this is still fine for review purposes, I'm still stacking one remaining test issue, which is that `TestSceneMultiplayerPlayer` will occasionally fail (both the ctor and the test method):

<img width="710" alt="JetBrains Rider-EAP 2022-03-17 at 18 28 26" src="https://user-images.githubusercontent.com/191335/158871279-1466e3be-e6cf-4b3a-9e04-140e83e387b1.png">

```csharp
TearDown : System.AggregateException : One or more errors occurred. (StartGameplay should not be called when the gameplay clock is already running)
  ----> System.InvalidOperationException : StartGameplay should not be called when the gameplay clock is already running
--TearDown
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at osu.Framework.Extensions.TaskExtensions.WaitSafely(Task task)
   at osu.Framework.Testing.TestScene.checkForErrors()
   at osu.Framework.Testing.TestScene.RunTestsFromNUnit()
--InvalidOperationException
   at osu.Game.Screens.Play.Player.StartGameplay() in /Users/dean/Projects/osu/osu.Game/Screens/Play/Player.cs:line 982
   at osu.Game.Screens.OnlinePlay.Multiplayer.MultiplayerPlayer.<onMatchStarted>b__22_0() in /Users/dean/Projects/osu/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:line 179
   at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Platform.HeadlessGameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.processFrame()
--- End of stack trace from previous location ---
   at osu.Framework.Platform.GameHost.<>c__DisplayClass126_0.<abortExecutionFromException>b__0()
   at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Platform.ThreadRunner.RunMainLoop()
   at osu.Framework.Platform.GameHost.windowUpdate()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Framework.Testing.TestScene.<>c__DisplayClass44_0.<SetupGameHostForNUnit>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)



-----

One or more child tests had errors
  Exception doesn't have a stacktrace



[performance] 2022-03-17 18:28:54 [verbose]: TextureAtlas initialised (1024x1024)
[runtime] 2022-03-17 18:28:54 [verbose]: 💨 Class: TestSceneMultiplayerPlayer
[runtime] 2022-03-17 18:28:54 [verbose]: 🔶 Test:  TestConstructor
[runtime] 2022-03-17 18:28:55 [verbose]: Opened realm "/var/folders/qt/mf5nyxrn07s0_mwpb_t8bl5h0000gn/T/of-test-headless/TestSceneMultiplayerPlayer-76193df2-95ab-493c-b558-e3eda2b8b72f/client.realm" at version 14
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #2 exit all screens
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #3 join room
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #4 wait for room join
[runtime] 2022-03-17 18:28:55 [verbose]: ✔️ 14 repetitions
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #5 set beatmap
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #6 initialise gameplay
[runtime] 2022-03-17 18:28:55 [verbose]: ScreenTestScene screen changed → MultiplayerPlayer
[runtime] 2022-03-17 18:28:55 [verbose]: 📺 ScreenTestScene(OsuScreenStack)#164(depth:1) loading MultiplayerPlayer#283
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #7 wait for player to be current
[performance] 2022-03-17 18:28:56 [verbose]: TextureAtlas size exceeded 1 time(s); generating new texture (1024x1024)
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 ScreenTestScene(OsuScreenStack)#164(depth:1) entered MultiplayerPlayer#283
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 BackgroundScreenStack#571(depth:1) loading BackgroundScreenBeatmap#603
[runtime] 2022-03-17 18:28:56 [verbose]: ✔️ 1383 repetitions
[runtime] 2022-03-17 18:28:56 [verbose]: 🔸 Step #8 start gameplay
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 BackgroundScreenStack#571(depth:1) entered BackgroundScreenBeatmap#603
[runtime] 2022-03-17 18:28:56 [error]: An unhandled error has occurred.
[runtime] 2022-03-17 18:28:56 [error]: System.InvalidOperationException: StartGameplay should not be called when the gameplay clock is already running
[runtime] 2022-03-17 18:28:56 [error]: at osu.Game.Screens.Play.Player.StartGameplay() in /Users/dean/Projects/osu/osu.Game/Screens/Play/Player.cs:line 982
[runtime] 2022-03-17 18:28:56 [error]: at osu.Game.Screens.OnlinePlay.Multiplayer.MultiplayerPlayer.<onMatchStarted>b__22_0() in /Users/dean/Projects/osu/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:line 179
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.Scheduler.Update()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
```

If someone beats me to figuring out why I will not complain.

peppy committed:
Add test coverage of multiplayer being paused when users are still loading

peppy committed:
Simplify clock reset/start flow in `MultiSpectatorScreen`

peppy committed:
Start `GameplayClockContainer` paused for better state control

peppy committed:
Move `DrawableRuleset` binding to `LoadComplete` to avoid exceptions on `InputManager` access

peppy committed:
Add more comments around catch up logic

peppy committed:
Fix test regression due to missing argument

peppy commented:
This is ready for review again ._.

peppy committed:
Fix `TestSceneMasterGameplayClockContainer` not clearing previous test clocks

frenzibyte commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -49,33 +49,38 @@ public class MasterGameplayClockContainer : GameplayClockContainer
         private readonly BindableDouble pauseFreqAdjust = new BindableDouble(1);
```
It's both funny and scary how something like this can cause tests to fail in a brain-melting way (definitely worth adding test coverage for):
```suggestion
        private readonly BindableDouble pauseFreqAdjust = new BindableDouble();
```
(`GameplayClockContainer` is now paused by default, so the adjustment should be `0`, otherwise will be stuck to `1` until `IsPaused` is toggled off and back on)

To explain why this is only failing when multiple test cases run in the same session (and I'm gonna pick the `TestSeekPerformsInGameplayTime(0.5d, -200.0d, true)` test case as an example while explaining): Since `pauseFreqAdjust` is set to 1 by default (which is incorrect), the first `ProcessFrame` call that the offset clocks receive are when `Reset` is called in the test case:
https://github.com/ppy/osu/blob/7cb7b03cee678db55530d8134d39295089a6a216/osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs#L86-L88

And, since another test has ran in the same session, `OsuSetting.AudioOffset` will already be populated with a certain offset (let's say -200ms) at the point of calling `Reset`, which ends up making `userGlobalOffsetClock.RateAdjustedOffset` matching that offset at the *current* rate (which is 1, not 0.5, since the track is recreated per-run, and the rate adjustment is applied in a later step).

Then, since on this test case the clock remains stopped, the second `ProcessFrame` call only happens during the `Seek(2500)` step, and the failure occurs there, since the following happens when `Seek` gets called:
 - The *current* `totalAppliedOffset` (which is going to be -200ms from above) is subtracted from the proposed time (2500ms), so the base seek call receives 2700ms
 - Seeking is performend and `GameplayClock.UnderlyingClock.ProcessFrame()` is called.
 - The `totalAppliedOffset` *changes* from -200ms to -100ms since this is the first time `ProcessFrame()` gets called while the track rate is actually 0.5 instead of 1.

This ends up making a difference in `CurrentTime` from the expected one since -200ms was subtracted from it at first but now `totalAppliedOffset` has been corrected to be -100ms instead (so it ends up with 2600ms instead of the proposed 2500ms).

---

With `pauseFreqAdjust` being changed to be 0 by default, `RateAdjustedOffset` will never be touched and tests will pass just fine.

While it may seem awkward that `HardwareCorrectionOffsetClock` only gets processed during seeks from the explaination above, this is only happening because of this edge case (`IsPaused = true && pauseFreqAdjust = 1`). If `IsPaused = false && pauseFreqAdjust = 1`, offsets will be updated per-frame, and if `IsPaused = true && pauseFreqAdjust = 0`, offsets will never be updated neither per-frame nor in seeks.

The only potential issue is when offsets are changed *and* the pause state changed without any prior frames processing the offset change, but I'm not sure that can be solveable anyways (and I might as well be proven wrong on this statement).

peppy replied:
What an ordeal. Thanks for the investigation. Will see what I can do with this knowledge.

peppy replied:
Applying your proposed change breaks other tests, so will need a bit of further thought:

![JetBrains Rider-EAP 2022-03-24 at 05 34 15](https://user-images.githubusercontent.com/191335/159849323-cf86a884-52c4-4199-a267-9bf7f2db0d9f.png)

peppy replied:
I've fixed the remaining tests and added some test coverage for this scenario. I'm still not 100% on this, but it's something to be considered at a later point. Specifically, I'm not sure how I feel about the gameplay clock potentially being in an incorrect state for the "current" frame because it's being updated as a child of `Player`.

At some point we will want to do an intensive consistency pass on the game loop in general to ensure we aren't inducing any single-frame delays as a result of having components processed too late.

peppy committed:
Merge branch 'master' into fix-spectator-seeks

peppy committed:
Set a more correct initial value for `pauseFreqAdjust`

As the `GameplayClock` now starts paused, this value needs to match to
ensure things work correctly.

For a better explanation of how we got here, see discussion at
https://github.com/ppy/osu/pull/17302#discussion_r830017735.

peppy committed:
Adjust `TestSceneFailAnimation` to account for initial frequency change

peppy committed:
Add test coverage of user offset being set before construction of `MasterGameplayClockContainer`

peppy commented:
[Two](https://github.com/ppy/osu/runs/5671658212?check_suite_focus=true) [of](https://github.com/ppy/osu/runs/5671658249?check_suite_focus=true) the test failures look like they may be related to these changes, but honestly I don't know how. Can't repro locally but I'll leave running in a hope to manage to.

peppy committed:
Merge branch 'master' into fix-spectator-seeks

peppy committed:
Add more comprehensive assert output to try and discern CI issues

peppy commented:
Test failures are related to the windows-specific platform offset being applied. Not yet sure on whether this is more or less correct. Maybe @frenzibyte will be able to come to a conclusion quicker than I can.

https://github.com/ppy/osu/runs/5846066167?check_suite_focus=true#step:6:1127

peppy committed:
Remove unused using statements

frenzibyte commented:
Hmm, the test failures raise a very good question on how `MasterGameplayClockContainer` should actually behave when initially stopped in regards to current time and offsets.

On master, since `pauseFreqAdjust = 1` by default, `CreateGameplayClock` would initialise offset clocks and update their `RateAdjustedOffset` as well. Making the first gameplay start time seek in `LoadComplete` work properly.

Meanwhile on this PR, since `pauseFreqAdjust = 0` by default, `CreateGameplayClock` would initialise offset clocks with `RateAdjustedOffset` remaining to be zero. Making the first gameplay start time seek in `LoadComplete` become off by the applied offsets (even though the gameplay clock is not paused at that point).

For a solution, I think that whenever the state of `IsPaused` changes, `pauseFreqAdjust` should update immediately and process the offset clocks to update the `RateAdjustedOffset`, so that the first gameplay start time seek happens properly with `RateAdjustedOffset` being populated:
```diff
diff --git a/osu.Game/Screens/Play/GameplayClockContainer.cs b/osu.Game/Screens/Play/GameplayClockContainer.cs
index 226f5709c7..f2df00d691 100644
--- a/osu.Game/Screens/Play/GameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/GameplayClockContainer.cs
@@ -122,14 +122,14 @@ public void Reset(bool startClock = false, double? gameplayStartTime = null)
             if (gameplayStartTime != null)
                 StartTime = gameplayStartTime;

-            ensureSourceClockSet();
-            Seek(StartTime ?? 0);
-
             // Manually stop the source in order to not affect the IsPaused state.
             AdjustableSource.Stop();

             if (!IsPaused.Value || startClock)
                 Start();
+
+            ensureSourceClockSet();
+            Seek(StartTime ?? 0);
         }

         /// <summary>
diff --git a/osu.Game/Screens/Play/MasterGameplayClockContainer.cs b/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
index f4445d3fa0..2802f721c8 100644
--- a/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
@@ -131,16 +131,29 @@ private double findEarliestStartTime()
         protected override void OnIsPausedChanged(ValueChangedEvent<bool> isPaused)
         {
             // The source is stopped by a frequency fade first.
-            if (isPaused.NewValue)
+            if (IsLoaded)
             {
-                this.TransformBindableTo(pauseFreqAdjust, 0, 200, Easing.Out).OnComplete(_ =>
+                if (isPaused.NewValue)
                 {
-                    if (IsPaused.Value == isPaused.NewValue)
-                        AdjustableSource.Stop();
-                });
+                    this.TransformBindableTo(pauseFreqAdjust, 0, 200, Easing.Out).OnComplete(_ =>
+                    {
+                        if (IsPaused.Value == isPaused.NewValue)
+                            AdjustableSource.Stop();
+                    });
+                }
+                else
+                    this.TransformBindableTo(pauseFreqAdjust, 1, 200, Easing.In);
             }
             else
-                this.TransformBindableTo(pauseFreqAdjust, 1, 200, Easing.In);
+            {
+                pauseFreqAdjust.Value = isPaused.NewValue ? 1 : 0;
+
+                if (isPaused.NewValue)
+                    AdjustableSource.Stop();
+
+                // process underlying gameplay clocks to update rate-adjusted offsets with the new frequency adjustment.
+                GameplayClock.UnderlyingClock.ProcessFrame();
+            }
         }

         public override void Start()
```

The above diff also reorders `Reset` such that it updates the running state of the clock before performing a seek, for the `RateAdjustedOffset` values to be updated properly for correct seeking.

peppy committed:
Ensure running state is updated before performing `Seek` in `GameplayClockContainer.Reset`

peppy committed:
Fix `MasterGameplayClockContainer` having incorrect rate-based offsets immediately after `LoadComplete`

I've attempted to explain why this is required using inline comments.
There's also further conversation at
https://github.com/ppy/osu/pull/17302#issuecomment-1091850927.

peppy commented:
I've applied your proposed fixes with some minor changes. Seems good, and makes sense 👍 .

peppy committed:
Remove unnecessary string literals

bdach commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -106,16 +115,21 @@ public virtual void Seek(double time)
         {
-            ensureSourceClockSet();
-            Seek(0);
+            if (gameplayStartTime != null)
+                StartTime = gameplayStartTime;
```
Not sure how to feel about this param and it being used like this, it's a weird edgecasey state mutation in a method that is supposed to be a "reset" method. Is it an option to remove it from `Reset()` and instead explicitly set `StartTime` in call sites that need it?

peppy replied:
I've made this change, I think it feels better.

smoogipoo replied:
Is this PR missing a commit? The parameter's still here.

peppy replied:
You might be looking at an old diff, it's definitely not there for me

bdach commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -46,36 +46,41 @@ public class MasterGameplayClockContainer : GameplayClockContainer
+        /// </summary>
+        /// <param name="beatmap">The beatmap to be used for time and metadata references.</param>
+        /// <param name="latestGameplayStartTime">The latest time which should be used when introducing gameplay. Will be used when skipping forward.</param>
+        /// <param name="startFromLatestStartTime">Whether to start from the provided latest start time rather than zero.</param>
```
I don't follow what this `startFromLatestStartTime` toggle is for. When would you want/not want this behaviour? I tried to infer from usages but it looks almost entirely random.

smoogipoo replied:
Only thing I can gather is it looks like making up for `gameplayStartTime` not being a `double?`... Definitely weird and xmldoc doesn't help.

Also not sure why there are 3 ways of providing the start time - via `MasterGameplayClock` ctor, via `StartTime`, and via `Reset()`.

peppy replied:
I'll look into the final parameter as I agree it's a bit weird and usages are hard to follow.

> Also not sure why there are 3 ways of providing the start time - via `MasterGameplayClock` ctor, via `StartTime`, and via `Reset()`.

The ctor is setting `latestGameplayStartTime`, which has a very different meaning and purpose. Can you double-check the xmldoc and confirm you can understand from it? The ctor one doesn't change post-construction and is used for cases like skipping forward to gameplay.

peppy replied:
How about `skipTargetTime`?

peppy replied:
> I don't follow what this `startFromLatestStartTime` toggle is for. When would you want/not want this behaviour? I tried to infer from usages but it looks almost entirely random.

I've removed this and updated usages. Hopefully the local usages should be more understandable due to `StartTime` being set directly.

smoogipoo replied:
New naming and xmldocs are much better.

bdach commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -607,30 +607,24 @@ private void performUserRequestedSkip()
         /// </summary>
         /// <remarks>
-        /// Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
+        /// This performance a non-frame-stable seek. Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
```
```suggestion
        /// This performs a non-frame-stable seek. Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
```

bdach reviewed (commented):
This is a surface level review, I haven't read any of the above, just read source. Nothing seems too bad, but there is one or two things I don't quite get.

smoogipoo commented on osu.Game/Screens/Play/MasterGameplayClockContainer.cs:
```diff
@@ -152,10 +183,10 @@ public override void Seek(double time)
         /// </summary>
         public void Skip()
         {
-            if (GameplayClock.CurrentTime > gameplayStartTime - MINIMUM_SKIP_TIME)
+            if (GameplayClock.CurrentTime > latestGameplayStartTime - MINIMUM_SKIP_TIME)
```
What's the difference in referring to `latestGameplayStartTime` here instead of `StartTime`?

peppy replied:
As mentioned in my last comment, they are completely different variables with different meanings and usages. Can you have a read over the code which uses `latestGameplayStartTime` and potentially suggest a better name for it?

peppy committed:
Fix typo in xmldoc

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy committed:
Merge branch 'master' into fix-spectator-seeks

peppy committed:
Expose and set `GameplayStartTime` directly, rather than via `Reset` parameter

peppy commented:
Apologies for force push, stuff got very mixed up (shouldn't affect review as it was done quickly).

peppy committed:
Rename `latestGameplayStartTime` to `skipTargetTime`

peppy committed:
Remove `startFromSkipTarget` parameter and update usages that required said behaviour

peppy commented on osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:
```diff
@@ -73,26 +73,29 @@ public void TestElapseThenReset()
             ClockBackedTestWorkingBeatmap working = null;
             GameplayClockContainer gameplayClockContainer = null;
 
+            if (setAudioOffsetBeforeConstruction)
+                AddStep($"preset audio offset to {userOffset}", () => localConfig.SetValue(OsuSetting.AudioOffset, userOffset));
```
I thought the use of `LocalStorage` here was enough. I'll double-check.

peppy replied:
This whole scene is marked `[HeadlessTest]`, so this shouldn't be an issue.

smoogipoo replied:
I did consider that this is a headless test, but it depends on the ordering of events. Like if nunit ever reorders tests to run an `== true` test before an `== false`, then it's going to be in the wrong state. Rider likes to reorder its tests so...

Guess we can roll with it if you disagree, it's more of a correctness thing.

peppy replied:
Fair, I've added a reset before each test run.

peppy committed:
Reset audio offset before each test run in `TestSceneMasterGameplayClockContainer`

smoogipoo committed:
Merge branch 'master' into fix-spectator-seeks

smoogipoo committed:
Cleanup whitespace

smoogipoo reviewed (approved):
Lgtm

smoogipoo commented:
Not sure if the `TestScenePause` failure is something we've seen before...

peppy commented:
Yeah I'm not sure. Unless anyone has bright ideas I'd probably go forward with merge and see how often it pops up...

bdach reviewed (approved):
I think this cleaned up rather nicely with the latest batch of commits :+1:
